### PR TITLE
US123706 Quizzes: Retake incorrect only siren entity

### PR DESCRIPTION
**Pull request by mostlyfabulous**
[US123706](https://rally1.rallydev.com/#/29180338367d/dashboard?detail=%2Fuserstory%2F478977336380%2Ftasks&fdp=true?fdp=true): [ui] Attempts: Conditions entities

### CODE REVIEW

File: src/activities/quizzes/attempts/QuizAttemptsEntity.js [hunk 1]
```diff
@@ -79,6 +89,22 @@ export class QuizAttemptsEntity extends Entity {
...
+		return this._entity && this._entity.getSubEntityByClass(Classes.quizzes.attempts.retakeIncorrectOnly);
+	}
+
+	/**
+	 * @returns {bool} is quiz retake incorrect only
+	 */
+	retakeIncorrectOnly() {
+		const entity = this.getRetakeIncorrectOnlySubEntity();
+		if (!entity) return;
+		return entity.hasClass(Classes.quizzes.retakeIncorrectOnly) && entity.hasClass('checked');
```
[the remark this answers is not in the document]
Comment (mostlyfabulous) [reply src/activities/quizzes/attempts/QuizAttemptsEntity.js:105]: Thanks for the catch!

File: src/activities/quizzes/attempts/QuizAttemptsEntity.js [hunk 2]
```diff
@@ -79,6 +89,22 @@ export class QuizAttemptsEntity extends Entity {
...
+	 * @returns {object} quiz retake incorrect only sub-entity
+	 */
+	getRetakeIncorrectOnlySubEntity() {
+		return this._entity && this._entity.getSubEntityByClass(Classes.quizzes.attempts.retakeIncorrectOnly);
+	}
+
+	/**
+	 * @returns {bool} is quiz retake incorrect only
+	 */
+	retakeIncorrectOnly() {
```
[the remark this answers is not in the document]
Comment (mostlyfabulous) [reply src/activities/quizzes/attempts/QuizAttemptsEntity.js:102]: Looking at the `QuizEntity` yes, I'll align that with the other bools.

**Review by ronan-f — APPROVED** [review]
LGTM

File: src/activities/quizzes/attempts/QuizAttemptsEntity.js [hunk 3]
```diff
@@ -105,7 +131,8 @@ export class QuizAttemptsEntity extends Entity {
 		if (!this.canUpdateAttemptsAllowed()) return;
 		const {action, fields} = this._generateAttemptsAllowedAction(attemptsAllowed) || {};
 		if (!action) return;
-		await performSirenAction(this._token, action, fields);
+		const returnedEntity = await performSirenAction(this._token, action, fields);
```
Comment (wongvincent) [inline src/activities/quizzes/attempts/QuizAttemptsEntity.js:134]: We should check that `returnedEntity` is a value and if it isn't, just return.

File: src/activities/quizzes/attempts/QuizAttemptsEntity.js [hunk 4]
```diff
@@ -134,6 +161,37 @@ export class QuizAttemptsEntity extends Entity {
 		if (!this.canUpdateOverallGradeCalculation()) return;
 		const {action, fields} = this._generateOverallGradeCalculationTypeAction(calculationType) || {};
 		if (!action) return;
-		await performSirenAction(this._token, action, fields);
+		const returnedEntity = await performSirenAction(this._token, action, fields);
+		return new QuizAttemptsEntity(returnedEntity);
```
Comment (wongvincent) [inline src/activities/quizzes/attempts/QuizAttemptsEntity.js:165]: Check that `returnedEntity` is a value

File: src/activities/quizzes/attempts/QuizAttemptsEntity.js [hunk 5]
```diff
@@ -134,6 +161,37 @@ export class QuizAttemptsEntity extends Entity {
...
+		return { action, fields };
+
+	}
+
+	async setRetakeIncorrectOnly(retakeIncorrectOnly) {
+		if (!retakeIncorrectOnly || !this._hasRetakeIncorrectOnlyChanged(retakeIncorrectOnly)) return;
+		if (!this.canUpdateOverallGradeCalculation()) return;
+		const {action, fields} = this._generateRetakeIncorrectOnlyAction(retakeIncorrectOnly) || {};
+		if (!action) return;
+		const returnedEntity = await performSirenAction(this._token, action, fields);
```
Comment (wongvincent) [inline src/activities/quizzes/attempts/QuizAttemptsEntity.js:194]: check if `returnedEntity` is a value

**Pull request merged**